src/pudl/analysis/record_linkage/embed_dataframe.py

```diff
@@ -190,6 +226,73 @@ def as_transformer(self):
         return FunctionTransformer(self.company_cleaner.apply_name_cleaning)
 
 
+class FuelTypeFiller(TransformStep):
+    """Fill missing fuel types from another column."""
+
+    name: str = "fuel_type_filler"
+    fuel_type_col: str = "fuel_type_code_pudl"
+    name_col: str = "plant_name"
+
+    def as_transformer(self):
+        """Return configured FuelTypeFiller."""
+        return FunctionTransformer(
+            _fill_fuel_type_from_name,
+            kw_args={"fuel_type_col": self.fuel_type_col, "name_col": self.name_col},
+        )
+
+
+def _extract_keyword_from_column(ser: pd.Series, keyword_list: list[str]) -> pd.Series:
+    """Extract keywords contained in a Pandas series with a regular expression."""
+    pattern = r"(?:^|\s+)(" + "|".join(keyword_list) + r")(?:\s+|$)"
+    return ser.str.extract(pattern, expand=False)
+
+
+def _fill_fuel_type_from_name(
+    df: pd.DataFrame, fuel_type_col: str, name_col: str
+) -> pd.DataFrame:
+    """Impute missing fuel type data from a name column.
+
+    If a missing fuel type code is contained in the plant name,
+    fill in the fuel type code PUDL for that record. E.g. "Washington Hydro"
+    """
+    if fuel_type_col not in df.columns:
+        raise AssertionError(f"{fuel_type_col} is not in dataframe columns.")
+    # TODO: dynamically get this list from core_eia__codes_energy_sources
```

Comment:
What is the best way to dynamically generate this fuel type list from `core_eia__codes_energy_sources`? I guess it could be a dagster op with that table as an input dependency, but that seems overly complicated. Generate it from the `fuel_type_code_pudl` column of the "core_eia__codes_energy_sources" table of `pudl.metadata.eia`? That seems cleaner. I guess it doesn't need to be dynamically generated but it would be nice if it was.

Reply:
It looks like you might be able to use `pudl.metadata.codes.CODE_METADATA["core_eia__codes_energy_sources"]`. This is used to set constraints from `fuel_type_code_pudl`, so I think all the information you need is there